aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h

```diff
@@ -400,15 +459,76 @@ namespace impl {
     static_assert(std::is_same<guts::typelist::typelist<ParameterTypes...>, typename guts::infer_function_traits_t<KernelFunctor>::parameter_types>::value,
       "Parameter types mismatch");
 
-    static ReturnType call(OperatorKernel* functor, ParameterTypes... args) {
+    static ReturnType call(OperatorKernel* functor, DispatchKeySet, ParameterTypes... args) {
       KernelFunctor* functor_ = static_cast<KernelFunctor*>(functor);
+      // We're explicitly taking in a dispatchKeySet and not forwarding it to the registered kernel.
+      // This functor is used by all kernels that do not expect to take in a DispatchKeySet,
+      // i.e. pretty much all manually written kernels
       return (*functor_)(std::forward<ParameterTypes>(args)...);
     }
   };
 
+  // This specialization is for kernels with at least one argument.
+  // It's only difference is an extra static assert to check that the first argument isn't a DispatchKeySet
+  template<class KernelFunctor, class ReturnType, class FirstParam, class... ParameterTypes>
+  struct wrap_kernel_functor_unboxed_<KernelFunctor, ReturnType(FirstParam, ParameterTypes...)> final {
+    static_assert(!std::is_same<FirstParam, DispatchKeySet>::value,
+      "wrap_kernel_functor_unboxed expects the first argument of the schema NOT to be a DispatchKeySet. "
+      "If your kernel signature's first argument is a DispatchKeySet (and you plan to compute dispatch keys "
+      "explicitly in your kernel), you should register it with impl_withKeys().");
+    static_assert(std::is_same<ReturnType, typename guts::infer_function_traits_t<KernelFunctor>::return_type>::value,
+      "Return type mismatch");
+    static_assert(std::is_same<guts::typelist::typelist<FirstParam, ParameterTypes...>, typename guts::infer_function_traits_t<KernelFunctor>::parameter_types>::value,
+      "Parameter types mismatch");
+
+    static ReturnType call(OperatorKernel* functor, DispatchKeySet, FirstParam firstArg, ParameterTypes... args) {
+      KernelFunctor* functor_ = static_cast<KernelFunctor*>(functor);
+      // Note [Plumbing Keys Through The Dispatcher 2]
+      // See Note [Plumbing Keys Through The Dispatcher] for the background.
+      // This functor explicitly takes in a dispatchKeySet and drops it on the floor- it does not forward it to the registered kernel.
+      //
+      // This is due to the calling convention within the dispatcher, which expects all registered kernels to have a first argument of type
+      // DispatchKeySet.
```

Comment:
Still stewing on ways we might avoid the junk arg - maybe by moving the template specialization upstream so instead of metaprogramming the functor to throw away the dispatch key for non-using kernels, we could partially specialize KernelFunction::call to extract and pass the dispatch key if the signature needed it

Reply:
Just to share the way I'm thinking about this: I think there's an upside and a downside to moving the template specialization upstream:

(+) we get rid of the junk arg earlier in the call sequence, so there are fewer opportunities for the execution to get slowed down by having to pass a DispatchKeySet from function to function

(-) It looks to me like pushing the specialization higher upstream requires adding specializations to every downstream function. In the example from your comment, I think that having two versions of `KernelFunction::call()` that do/don't throw away the DispatchKeySet would also require two versions of `BoxedKernelWrapper::call()` that do/don't take in a DispatchKeySet. That's because the DispatchKeySet argument gets special treatment and isn't pushed/popped from the stack (for perf reasons + because we don't want jit to know about it).

In another comment, Sebastian also pointed out that moving the specialization further **downstream** could reduce the code duplication even further: If I shove a `wrap_kernel_functor_unboxed` wrapper underneath `make_boxed_from_unboxed`, I can eliminate all of the code duplication in `make_boxed_from_unboxed`, since that function can now treat every kernel as if it takes in a DispatchKeySet and pass off the responsibility of when to throw it away to `wrap_kernel_functor_unboxed`.

That also means that it has a chance of hurting performance (if we're carrying around an extra DispatchKeySet through more functions), but as long as functions are getting inlined properly then we shouldn't see a perf hit (which I'm going to try and check).

This does seem like it's worth discussing though: not every function in the Dispatcher call-path is currently inlined, and each function that isn't inlined translates to an extra small perf hit in this PR. And inlining everything isn't necessarily a clear win- so maybe some extra duplication is worth a small perf hit.